src/Stratis.SmartContracts.Executor.Reflection/Contract.cs

```diff
@@ -0,0 +1,176 @@
+
+            // This should not happen without setting the appropriate binding flags
+            if (methodToInvoke.IsConstructor)
+                return ContractInvocationResult.Failure(ContractInvocationErrorType.MethodIsConstructor);
+
+            // This should not happen without setting the appropriate binding flags
+            if (methodToInvoke.IsPrivate)
+                return ContractInvocationResult.Failure(ContractInvocationErrorType.MethodIsPrivate);
+
+            // Restore the state of the instance
+            if(!this.initialized)
+                SetStateFields(this.instance, this.State);
+
+            return this.InvokeInternal(methodToInvoke, invokeParams);
+        }
+
+        /// <summary>
+        /// Shared internal logic for invoking a method.
+        /// </summary>
+        private IContractInvocationResult InvokeInternal(MethodBase method, object[] parameters)
+        {
+            try
+            {
+                object result = method.Invoke(this.instance, parameters);
+
+                return ContractInvocationResult.Success(result);
+            }
+            catch (ArgumentException argumentException)
+            {
+                // Parameters do not match
+                // This should not happen
+                return ContractInvocationResult.Failure(ContractInvocationErrorType.ParameterTypesDontMatch, argumentException);
+            }
+            catch (TargetInvocationException targetException) when (!(targetException.InnerException is OutOfGasException))
+            {
+                // Method threw an exception that was not an OutOfGasException
+                return ContractInvocationResult.Failure(ContractInvocationErrorType.MethodThrewException, targetException);
+            }
+            catch (TargetInvocationException targetException) when (targetException.InnerException is OutOfGasException)
+            {
+                // Method threw an exception that was an OutOfGasException
+                return ContractInvocationResult.Failure(ContractInvocationErrorType.OutOfGas);
+            }
+            catch (TargetParameterCountException parameterException)
+            {
+                // Parameter count incorrect
+                // This should not happen
+                return ContractInvocationResult.Failure(ContractInvocationErrorType.ParameterCountIncorrect, parameterException);
+            }            
+        }
+
+        /// <summary>
+        /// Uses reflection to set the state fields on the contract object.
+        /// </summary>
+        private static void SetStateFields(SmartContract smartContract, ISmartContractState contractState)
+        {
+            FieldInfo[] fields = typeof(SmartContract).GetFields(BindingFlags.Instance | BindingFlags.NonPublic);
+
+            foreach (FieldInfo field in fields)
+            {
+                switch (field.Name)
+                {
+                    case "gasMeter":
```

Comment:
Would be nice to have some kinda nameof thing here but could be impossible.

Not blocking cos same in restorer at the moment.

Reply:
Possible we could even just a single field to improve performance.